test/Events.Test/Controllers/CollectControllerTests.cs

```diff
@@ -0,0 +1,715 @@
+    }
+
+    [Theory]
+    [AutoData]
+    public async Task Post_MixedEventTypes_ProcessesAllEvents(
+        Guid userId, Guid cipherId, Guid orgId, CipherDetails cipherDetails, Organization organization)
+    {
+        _currentContext.UserId.Returns(userId);
+        cipherDetails.Id = cipherId;
+        organization.Id = orgId;
+        _cipherRepository.GetByIdAsync(cipherId, userId).Returns(cipherDetails);
+        _organizationRepository.GetByIdAsync(orgId).Returns(organization);
+        var events = new List<EventModel>
+        {
+            new EventModel
+            {
+                Type = EventType.User_ClientExportedVault,
+                Date = DateTime.UtcNow
+            },
+            new EventModel
+            {
+                Type = EventType.Cipher_ClientAutofilled,
+                CipherId = cipherId,
+                Date = DateTime.UtcNow
+            },
+            new EventModel
+            {
+                Type = EventType.Organization_ClientExportedVault,
+                OrganizationId = orgId,
+                Date = DateTime.UtcNow
+            }
+        };
+
+        var result = await _sut.Post(events);
+
+        Assert.IsType<OkResult>(result);
+        await _eventService.Received(1).LogUserEventAsync(userId, EventType.User_ClientExportedVault, Arg.Any<DateTime?>());
+        await _eventService.Received(1).LogCipherEventsAsync(
+            Arg.Is<IEnumerable<Tuple<Cipher, EventType, DateTime?>>>(tuples => tuples.Count() == 1)
+        );
+        await _eventService.Received(1).LogOrganizationEventAsync(organization, EventType.Organization_ClientExportedVault, Arg.Any<DateTime?>());
+    }
+
+    [Theory]
+    [AutoData]
+    public async Task Post_MoreThan50CipherEvents_LogsInBatches(Guid userId, List<CipherDetails> ciphers)
+    {
+        _currentContext.UserId.Returns(userId);
+        var events = new List<EventModel>();
+
+        for (int i = 0; i < 100; i++)
+        {
+            var cipher = ciphers[i % ciphers.Count];
+            _cipherRepository.GetByIdAsync(cipher.Id, userId).Returns(cipher);
+            events.Add(new EventModel
+            {
+                Type = EventType.Cipher_ClientAutofilled,
+                CipherId = cipher.Id,
+                Date = DateTime.UtcNow
+            });
+        }
+
+        var result = await _sut.Post(events);
+
+        Assert.IsType<OkResult>(result);
+        await _eventService.Received(2).LogCipherEventsAsync(
+            Arg.Is<IEnumerable<Tuple<Cipher, EventType, DateTime?>>>(tuples => tuples.Count() == 50)
+        );
+    }
+
+    [Theory]
+    [AutoData]
+    public async Task Post_Exactly50CipherEvents_LogsInSingleBatch(Guid userId, List<CipherDetails> ciphers)
+    {
+        _currentContext.UserId.Returns(userId);
+        var events = new List<EventModel>();
+
+        for (int i = 0; i < 50; i++)
+        {
+            var cipher = ciphers[i % ciphers.Count];
```

Comment:
consider using a foreach instead of a for loop to reduce complexity.

```
        foreach (var cipher in ciphers)
        {
            _cipherRepository.GetByIdAsync(cipher.Id, userId).Returns(cipher);
            events.Add(new EventModel
            {
                Type = EventType.Cipher_ClientAutofilled,
                CipherId = cipher.Id,
                Date = DateTime.UtcNow
            });
        }

```